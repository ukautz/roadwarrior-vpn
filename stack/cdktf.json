{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "74ca6711-f045-4875-b7ed-edf848e011f7",
  "terraformProviders": [
    {
      "name": "digitalocean",
      "version": "~>2.16",
      "source": "digitalocean/digitalocean"
    },
    {
      "name": "hcloud",
      "version": ">=1.32,<2",
      "source": "hetznercloud/hcloud"
    },
    "cloudinit@>=2.2,<3",
    "local"
  ],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
