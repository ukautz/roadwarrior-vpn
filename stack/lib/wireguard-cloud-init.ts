import { Construct } from "constructs";
import * as cloudinit from "../.gen/providers/cloudinit";

const defaultServerAddress = "192.168.6.1/24";
const defaultServerPort = "51397";
const defaultClientAddress = "192.168.6.10/24";

export interface WireguardCloudInitProps {
  vpnServerAddress?: string;
  vpnServerPort?: string;
  vpnServerPrivateKey: string;
  vpnClientPublicKey: string;
  vpnClientAddress?: string;
  base64Encode?: boolean;
}

export class WireguardCloudInit extends cloudinit.Config {
  constructor(scope: Construct, name: string, props: WireguardCloudInitProps) {
    super(scope, name, {
      base64Encode: props.base64Encode ?? true,
      gzip: props.base64Encode ?? true,
      part: [
        {
          contentType: "text/cloud-config",
          content: `#cloud-config
apt_upgrade: true
packages:
  - wireguard`,
        },
        {
          contentType: "text/x-shellscript",
          content: `#!/bin/bash

mkdir -p /etc/wireguard
cd /etc/wireguard

umask 077

cat > wg0.conf <<- EOF
[Interface]
Address = ${props.vpnServerAddress ?? defaultServerAddress}
ListenPort = ${props.vpnServerPort ?? defaultServerPort}
PrivateKey = ${props.vpnServerPrivateKey}
PostUp = iptables -A FORWARD -i %i -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
PostDown = iptables -D FORWARD -i %i -j ACCEPT; iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE

[Peer]
PublicKey = ${props.vpnClientPublicKey}
AllowedIPs = ${props.vpnClientAddress ?? defaultClientAddress}
EOF

umask 022
sysctl -w net.ipv4.ip_forward=1
echo "net.ipv4.ip_forward = 1" >> /etc/sysctl.d/forwarding.conf

systemctl enable wg-quick@wg0
systemctl start wg-quick@wg0`,
        },
      ],
    });
  }
}
