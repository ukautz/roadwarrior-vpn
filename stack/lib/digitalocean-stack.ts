import { TerraformOutput, TerraformStack } from "cdktf";
import * as cloudinit from "../.gen/providers/cloudinit";
import * as digitalocean from "../.gen/providers/digitalocean";
import * as fs from "fs";
import {
  WireguardCloudInit,
  WireguardCloudInitProps,
} from "./wireguard-cloud-init";
import { Construct } from "constructs";

const defaultServerImage = "ubuntu-20-04-x64";
const defaultServerSize = "s-1vcpu-1gb";
const defaultServerRegion = "fra1";

export interface DigitalOceanStackProps extends WireguardCloudInitProps {
  digitalOceanToken: string;
  sshKeyPath: string;
  serverImage?: string;
  serverSize?: string;
  serverRegion?: string;
}

export class DigitalOceanStack extends TerraformStack {
  constructor(scope: Construct, name: string, props: DigitalOceanStackProps) {
    super(scope, name);

    new digitalocean.DigitaloceanProvider(this, "CloudProvider", {
      token: props.digitalOceanToken,
    });

    new cloudinit.CloudinitProvider(this, "CloudInitProvider");

    const publicKey = fs.readFileSync(props.sshKeyPath, "utf8");
    const sshKey = new digitalocean.SshKey(this, "SshKey", {
      name: `${name} Key`,
      publicKey,
    });

    const userData = new WireguardCloudInit(this, "WireguardCloudInit", {
      ...props,
      base64Encode: false,
    });

    const server = new digitalocean.Droplet(this, "VpnServer", {
      name: "vpn-server",
      image: props.serverImage ?? defaultServerImage,
      size: props.serverSize ?? defaultServerSize,
      region: props.serverRegion ?? defaultServerRegion,
      sshKeys: [sshKey.id],
      userData: userData.rendered,
    });

    [
      ["server-id", server.id],
      ["server-ip", server.ipv4Address],
      ["server-status", server.status],
      ["server-vpn-port", userData.serverPort],
      ["server-vpn-network", userData.serverAddress],
      ["client-vpn-address", userData.clientAddress],
    ].map(
      ([k, v]) =>
        new TerraformOutput(this, k as string, {
          value: v,
        })
    );
  }
}
