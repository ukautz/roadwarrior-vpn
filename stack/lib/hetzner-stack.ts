import { TerraformOutput, TerraformStack } from "cdktf";
import * as cloudinit from "../.gen/providers/cloudinit";
import * as hcloud from "../.gen/providers/hcloud";
import * as fs from "fs";
import {
  WireguardCloudInit,
  WireguardServerProps,
} from "./wireguard-cloud-init";
import { Construct } from "constructs";

const defaultServerImage = "ubuntu-20.04";
const defaultServerType = "cx11";
const defaultServerLocation = "nbg1";

export interface HetznerStackProps extends WireguardServerProps {
  hcloudToken: string;
  sshKeyPath: string;
  serverImage?: string;
  serverType?: string;
  serverLocation?: string;
}

export class HetznerStack extends TerraformStack {
  constructor(scope: Construct, name: string, props: HetznerStackProps) {
    super(scope, name);

    new hcloud.HcloudProvider(this, "CloudProvider", {
      token: props.hcloudToken,
    });

    new cloudinit.CloudinitProvider(this, "CloudInitProvider");

    const publicKey = fs.readFileSync(props.sshKeyPath, "utf8");
    const sshKey = new hcloud.SshKeyA(this, "SshKey", {
      name: `${name} Key`,
      publicKey,
    });

    const userData = new WireguardCloudInit(this, "WireguardCloudInit", {
      ...props,
      sshKey: publicKey,
      base64Encode: true,
    });

    const server = new hcloud.Server(this, "VpnServer", {
      name: "vpn-server",
      image: props.serverImage ?? defaultServerImage,
      serverType: props.serverType ?? defaultServerType,
      location: props.serverLocation ?? defaultServerLocation,
      sshKeys: [sshKey.id],
      userData: userData.rendered,
    });

    [
      ["server-id", server.id],
      ["server-ip", server.ipv4Address],
      ["server-status", server.status],
      ["server-vpn-port", userData.serverPort],
      ["server-vpn-network", userData.serverAddress],
      ["client-vpn-address", userData.clientAddress],
      ["ssh-user", userData.sshUser],
    ].map(
      ([k, v]) =>
        new TerraformOutput(this, k as string, {
          value: v,
        })
    );
  }
}
