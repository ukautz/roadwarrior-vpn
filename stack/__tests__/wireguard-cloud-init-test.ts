import { Testing } from "cdktf";
import "cdktf/lib/testing/adapters/jest"; // Load types for expect matchers
import { WireguardCloudInit } from "../lib/wireguard-cloud-init";

describe("Wireguard Cloud Init", () => {
  it("Renders as expected", () => {
    expect(
      Testing.synthScope((scope) => {
        new WireguardCloudInit(scope, "Resource", {
          vpnClientAddress: "10.0.0.10/24",
          vpnClientPublicKey: "pub-key",
          vpnServerAddress: "10.0.0.1/24",
          vpnServerPort: "12345",
          vpnServerPrivateKey: "priv-key",
        });
      })
    ).toMatchSnapshot();
  });
});
