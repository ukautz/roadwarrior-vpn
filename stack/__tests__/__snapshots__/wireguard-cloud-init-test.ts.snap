// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wireguard Cloud Init Renders as expected 1`] = `
"{
  \\"resource\\": {
    \\"cloudinit_config\\": {
      \\"Resource\\": {
        \\"base64_encode\\": true,
        \\"gzip\\": true,
        \\"part\\": [
          {
            \\"content\\": \\"#cloud-config\\\\napt_upgrade: true\\\\npackages:\\\\n  - wireguard\\",
            \\"content_type\\": \\"text/cloud-config\\"
          },
          {
            \\"content\\": \\"#!/bin/bash\\\\n\\\\nmkdir -p /etc/wireguard\\\\ncd /etc/wireguard\\\\n\\\\numask 077\\\\n\\\\ncat > wg0.conf <<- EOF\\\\n[Interface]\\\\nAddress = 10.0.0.1/24\\\\nListenPort = 12345\\\\nPrivateKey = priv-key\\\\nPostUp = iptables -A FORWARD -i %i -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE\\\\nPostDown = iptables -D FORWARD -i %i -j ACCEPT; iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE\\\\n\\\\n[Peer]\\\\nPublicKey = pub-key\\\\nAllowedIPs = 10.0.0.10/24\\\\nEOF\\\\n\\\\numask 022\\\\nsysctl -w net.ipv4.ip_forward=1\\\\necho \\\\\\"net.ipv4.ip_forward = 1\\\\\\" >> /etc/sysctl.d/forwarding.conf\\\\n\\\\nsystemctl enable wg-quick@wg0\\\\nsystemctl start wg-quick@wg0\\",
            \\"content_type\\": \\"text/x-shellscript\\"
          }
        ]
      }
    }
  }
}"
`;
